### ISSUE TITLE
data corruption with multi-threading and copy-on-write enabled

### ISSUE THREAD

**Issue by stephane-chazelas** [NetworkBlockDevice/nbd#43]
Since the introduction of multithreading, it looks like writing to a nbd device can corrupt data as the data is not written to right offset by nbd-server. Setting max_threads to 1 works around it, but I still ran into deadlocks. Details below.

I reproduced it consistently using cow with a server on Cygwin 32bit or Linux 32 bit, but I suspect that it can also affect 64 bit and non cow exports.

I initially ran into the problem when exporting a VSS snapshot of a disk partition with cow from Cygwin. Connecting the nbd device, mount the NTFS file system on it, making modifications, flush data to the nbd device, and the NTFS file system ends up corrupted.

I can reproduce the corruption easily (more easily in 32 bits for some reason) with (with the server on Linux as well) by doing:

```
truncate -s 500G file # 500GiB sparse file
cat > conf << \EOF
[generic]
        port = 12345
#       max_threads = 1

[default]
        exportname = /path/to/file
        copyonwrite = true
        maxconnections = 1
EOF
nbd-server -C conf
```

And connect that on the client and do:

```
printf abc | sudo dd bs=4k seek=10 conv=notrunc of=/dev/nbd0
printf xyz | sudo dd bs=4k seek=100 conv=notrunc of=/dev/nbd0
printf XYZ | sudo dd bs=4k seek=1000 conv=notrunc of=/dev/nbd0
blockdev --flushbufs /dev/nbd0
```

I found the diff file had:

```
0x0000 all zeros
0x1000 3rd new block
0x2000 1st new block
```

When compiling the server with debug, I see:

```
*handling read request
Asked to read 1024 bytes at 40960.
Page 10 is not here, we read the original one
(READ from fd 6 offset 40960 len 1024), ++

*handling read request
Asked to read 1024 bytes at 409600.
Page 100 is not here, we read the original one
(READ from fd 6 offset 409600 len 1024), ++

****handling write request
Asked to write 1024 bytes at 409600.
handling write request
Asked to write 1024 bytes at 40960.
Page 10 is not here, we put it at 1
(READ from fd 6 offset 40960 len 4096), Page 100 is not here, we put it at 0
(READ from fd 6 offset 409600 len 4096),++*handling read request
Asked to read 1024 bytes at 4096000.
Page 1000 is not here, we read the original one
(READ from fd 6 offset 4096000 len 1024), ++

**handling write request
Asked to write 1024 bytes at 4096000.
Page 1000 is not here, we put it at 2
(READ from fd 6 offset 4096000 len 4096), +
```

Look how the intermingled output suggests the requests are being processed in different threads.

What I think is happening is that each of the threads does a lseek() followed by a write(). But because there's no exclusive lock, what can happen is that the order of processing, instead of being `thread1.lseek();thread1.write();thread2.lseek();thread2.write()`,  is instead: `thread1.lseek();thread2.lseek();thread1.write();thread2.write()` and the data ends up being written in the wrong place.

With `max_threads = 1`, I no longer see those problems, but I see some occasional deadlocks (in my cases when doing a ntfsclone on the nbd device). With max_threads = 1, we still have a thread handling the requests and another working on them. I suspect there's still some shared state/memory that is not properly mutexed.

Using `pwrite()` instead of `lseek()`+`write()` may remove the need for exclusive lock although I don't know if one gets a guarantee of atomicity regardless of the size of that data being written.

I ended up reverted to 3.9.1 where it runs as expected.

**Comment by abligh** [comment]
Yuck. That's a pretty serious issue. Thanks for finding this.

Yes, `pwrite` (and no doubt the same for read) is the right fix here.

**Comment by yoe** [comment]
the other writes already use `pread` and `pwrite`, but apparently copyonwrite does not. Obviously that does need to be fixed, indeed.

**Comment by stephane-chazelas** [comment]
Note that the `pread`/`pwrite` would fix the corruption issue but probably not the deadlock issue. It may be worth logging that as a separate issue. I didn't investigate that one much as reverting to 3.9.1 worked for me and also fixed several other issues I have with newer versions like the fact that it's incompatible with xnbd, qemu-nbd or nbdkit or that it's cumbersome to use as you need to create a temporary configuration file for a one-of export.

For the record, the deadlock (with max_threads = 1) was occurring at random in the middle of a ntfsclone (from nbd device on Cygwin to a raw file on Linux, 3.14 for both the server and client). I've not done any test with the server running on Linux, so maybe it's just Cygwin/Windows pthread implementation that's buggy.

**Comment by yoe** [comment]
On Thu, Dec 22, 2016 at 03:03:47AM -0800, Stephane Chazelas wrote:
> Note that the pread/pwrite would fix the corruption issue but probably not the
> deadlock issue. It may be worth logging that as a separate issue.

Yes please.

> I didn't investigate that one much as reverting to 3.9.1 worked for me and
> also fixed several other issues I have with newer versions like the fact that
> it's incompatible with xnbd,

How?

> qemu-nbd

How?

> or nbdkit

How?

> or that it's cumbersome to use as you need to create a temporary
> configuration file for a one-of export.

This was fixed in commit 25fc7b3c, nbd 3.12.

> For the record, the deadlock (with max_threads = 1) was occurring at random in
> the middle of a ntfsclone (from nbd device on Cygwin to a raw file on Linux,
> 3.14 for both the server and client). I've not done any test with the server
> running on Linux, so maybe it's just Cygwin/Windows pthread implementation
> that's buggy.

That, or it could be the client, or it could be ntfsclone that doesn't
expect things to happen in the way they do.

I don't see deadlocks in my case, so if you want that to be an
actionable bug report you're going to have to give me a lot more
information.

I've got a preliminary fix for the copyonwrite issue, but it still needs
finishing. Meanwhile max_threads = 1 should be a reasonable workaround.

-- 
< ron> I mean, the main *practical* problem with C++, is there's like a dozen
       people in the world who think they really understand all of its rules,
       and pretty much all of them are just lying to themselves too.
 -- #debian-devel, OFTC, 2016-02-12

**Comment by stephane-chazelas** [comment]
2016-12-22 04:34:19 -0800, Wouter Verhelst:
[...]
> > I didn't investigate that one much as reverting to 3.9.1 worked for me and
> > also fixed several other issues I have with newer versions like the fact that
> > it's incompatible with xnbd,
> 
> How?
> 
> > qemu-nbd
> 
> How?
> 
> > or nbdkit
> 
> How?

AFAIK, none of those support the new protocol and nbd-client no
longer supports the old protocol (BTW, I've also seen
incompatibilities between versions of nbd (like a writable
export by nbd-server 3.14 appears read-only when connected with
a nbd-client 3.7 (as found on ubuntu 14.04))).

> > or that it's cumbersome to use as you need to create a temporary
> > configuration file for a one-of export.
> 
> This was fixed in commit 25fc7b3c, nbd 3.12.

Thanks. I missed that.

[...]
> I don't see deadlocks in my case, so if you want that to be an
> actionable bug report you're going to have to give me a lot more
> information.
[...]

Yes, I'll raise a separate issue once I've time to try and
reproduce it.

Cheers,
Stephane

**Comment by yoe** [comment]
On Thu, Dec 22, 2016 at 04:58:15AM -0800, Stephane Chazelas wrote:
> 2016-12-22 04:34:19 -0800, Wouter Verhelst:
> [...]
> > > I didn't investigate that one much as reverting to 3.9.1 worked for me and
> > > also fixed several other issues I have with newer versions like the fact
> that
> > > it's incompatible with xnbd,
> >
> > How?
> >
> > > qemu-nbd
> >
> > How?
> >
> > > or nbdkit
> >
> > How?
> 
> AFAIK, none of those support the new protocol

qemu-nbd and nbdkit have since been updated to support the new protocol
(qemu already did, and it used a common implementation with just the
negotiation differing, so there the patch was trivially easy)

xnbd has always supported newstyle. However, xnbd has a few features
that its designer chose not to document, and therefore that feature is
incompatible with nbd. Not much we can do about that (other than reading
a foreign code base and, against the author's express wishes,
reimplement the feature).

IOW, your claims are just wrong. Please don't throw unsubstanciated
claims around like that. We've spent a *lot* of time ensuring the
several implementations are compatible with one another; what you're
doing is spreading rumours that aren't based on reality, and then we
will have to rebut them for a very long time. That's not nice.

> and nbd-client no longer supports the old protocol

Indeed, after five years it was time to drop it.

> (BTW, I've also seen incompatibilities between versions of nbd (like a
> writable export by nbd-server 3.14 appears read-only when connected with a
> nbd-client 3.7 (as found on ubuntu 14.04))).

Yes, that was an unfortunate forward-compatibility bug in nbd-client.
Debian 8.7 will have a fix for that (it's already in
stable-proposed-updates), and I believe Alex was working on an update
for (some of) Ubuntu's LTS releases as well. Alex?

> > > or that it's cumbersome to use as you need to create a temporary
> > > configuration file for a one-of export.
> >
> > This was fixed in commit 25fc7b3c, nbd 3.12.
> 
> Thanks. I missed that.

Not the only thing, apparently ;)

-- 
< ron> I mean, the main *practical* problem with C++, is there's like a dozen
       people in the world who think they really understand all of its rules,
       and pretty much all of them are just lying to themselves too.
 -- #debian-devel, OFTC, 2016-02-12

**Comment by abligh** [comment]
Re Ubuntu, it's https://bugs.launchpad.net/ubuntu/trusty/+source/nbd/+bug/1578185

rbasak was going to do an SRU for this. I shall ping him.

**Comment by stephane-chazelas** [comment]
2016-12-22 05:30:56 -0800, Wouter Verhelst:
[...]
> qemu-nbd and nbdkit have since been updated to support the new protocol
> (qemu already did, and it used a common implementation with just the
> negotiation differing, so there the patch was trivially easy)
> 
> xnbd has always supported newstyle. However, xnbd has a few features
> that its designer chose not to document, and therefore that feature is
> incompatible with nbd. Not much we can do about that (other than reading
> a foreign code base and, against the author's express wishes,
> reimplement the feature).
[...]

Sorry, my bad for xnbd and for not checking with the latest
versions with qemu and nbdkit. That boils down to me getting
frustrated as I couldn't access the exports of nbdkit and
qemu-nbd from /older/ systems after an upgrade of nbd on one of my
machines.

Looks like the qemu-nbd and nbdkit shipped with Ubuntu 16.04 at
least don't support the new protocol yet, while 16.04 comes with
nbd 3.13 which I beleive doesn't support the old one. It feels like
you could have waited a little longer before droppig support for the
old style, or Ubuntu should not have upgraded nbd, when the
other tools didn't support the new protocol yet.

-- 
Stephane

**Comment by abligh** [comment]
Stephane,

The protocol did not change between the two versions (well not in a manner that caused the bug). What happened is 3.13 exposed a new flag, and a bug in the previous version (3.9 here, but it was introduced before) caused this flag to be sent incorrectly to the kernel, in a bit position that happens to map to read-only. The problem was a bug in the client in previous versions, which could have been exposed by *any* NBD server (and indeed was - there was a gonbdserver issue raised, and it was originally discovered as a qemu nbd server issue).

Alex

**Comment by abligh** [comment]
Branch 3.7-fix-flags, i.e.
  https://github.com/NetworkBlockDevice/nbd/tree/3.7-fix-flags
should fix the read-only issue, specifically this commit: https://github.com/NetworkBlockDevice/nbd/commit/9b591289014b50d41349bd684c80443315513c71

Want to give that a go? If it works I'll ask rbasak to do an SRU (stable release update for Ubuntu).

**Comment by yoe** [comment]
@stephane-chazelas, I just pushed what I hope will be a fix. Can you try if you can reproduce it with that patch applied?

**Comment by abligh** [comment]
You know what @yoe , we should really make the integrity test run on treefiles too, which might well have caught this.

**Comment by yoe** [comment]
@abligh, treefiles isn't affected, so it wouldn't :-)

We already have a test for copyonwrite, but it doesn't run the integrity test, so it doesn't check for data corruption. We should indeed fix that up, though.

I'm currently a bit swamped with other things, but if you have the time, feel free...

**Comment by abligh** [comment]
@yoe OK I'll see if I can get to that this week. I may refactor the testing stuff entirely as it's got a bit duplicative.

**Comment by yoe** [comment]
@abligh no contest there. Much of the huge switch is just copy/paste code. Would be nice to see that cleaned up :)

**Comment by yoe** [comment]
This is now also filed in the Debian BTS as [852288](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=852288)

**Comment by yoe** [comment]
this is now fixed in a released version of nbd (3.15.2)
